mbtiles/__init__.py

```diff
@@ -47,19 +50,35 @@ def process_tile(tile):
     kwds['transform'] = from_bounds(ulx, lry, lrx, uly, 256, 256)
     src_nodata = kwds.pop('src_nodata', None)
     dst_nodata = kwds.pop('dst_nodata', None)
-
-    with rasterio.open('/vsimem/tileimg', 'w', **kwds) as tmp:
-        reproject(rasterio.band(src, src.indexes),
-                  rasterio.band(tmp, tmp.indexes),
-                  src_nodata=src_nodata,
-                  dst_nodata=dst_nodata,
-                  num_threads=1,
-                  resampling=resampling)
-
-    data = bytearray(virtual_file_to_buffer('/vsimem/tileimg'))
-
-    # Workaround for https://bugs.python.org/issue23349.
-    if sys.version_info[0] == 2 and sys.version_info[2] < 10:
-        data[:] = data[-1:] + data[:-1]
-
-    return tile, data
+    kwds['driver'] = 'GTiff'
+
+    # Reproject into a GeoTIFF tile.
+    with MemoryFile() as memdst:
+        with memdst.open(**kwds) as tiledst:
+            reproject(
+                rasterio.band(src, src.indexes),
+                rasterio.band(tiledst, tiledst.indexes),
+                src_nodata=src_nodata, dst_nodata=dst_nodata, num_threads=2,
+                resampling=resampling)
+
+        # If the tiledst dataset has any valid data, we read and return it.
+        # Otherwise, we return None.
+        #
+        # To save time, we check the only mask of the first band for
+        # valid data.
+        with memdst.open() as tiledst:
+            mask = tiledst.read_masks(1)
+            if mask.any():
+                # We're using MemoryFile to get a filename from
+                # the in-memory filesystem only; it doesn't support
+                # copy yet.
+                with MemoryFile() as memimg:
+                    copy(tiledst, memimg.name, driver=base_kwds['driver'])
```

Comment:
This new version of copy requires rasterio>=1.0a11

Reply:
@jdmcbr I'm almost certain that the reason #36 is slower than #34 is the extra copy here at line 76.